From f2831991561afc3cad808edf1ecde1a02fa4ca58 Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@cavium.com>
Date: Mon, 23 May 2016 14:21:36 -0700
Subject: [PATCH 32/39] Revert "net: thunderx: Fix BGX's QSGMII mode
 configuration"

Some unwanted changes have crept in, will commit properly

This reverts commit 374e9888a60bc20836a4bdd7dc78d2b9c85b5e8d.

Signed-off-by: Sunil Goutham <sgoutham@cavium.com>
---
 drivers/net/ethernet/cavium/thunder/thunder_bgx.c | 72 +++++------------------
 1 file changed, 14 insertions(+), 58 deletions(-)

diff --git a/drivers/net/ethernet/cavium/thunder/thunder_bgx.c b/drivers/net/ethernet/cavium/thunder/thunder_bgx.c
index 470df9b..c8b96b8 100644
--- a/drivers/net/ethernet/cavium/thunder/thunder_bgx.c
+++ b/drivers/net/ethernet/cavium/thunder/thunder_bgx.c
@@ -137,18 +137,10 @@ EXPORT_SYMBOL(bgx_get_map);
 int bgx_get_lmac_count(int node, int bgx_idx)
 {
 	struct bgx *bgx;
-	struct lmac *lmac;
 
 	bgx = bgx_vnic[(node * MAX_BGX_PER_CN88XX) + bgx_idx];
-	if (bgx) {
-		/* For QSGMII although all 4 LMACs are enabled,
-		 * there is only one physical interface.
-		 */
-		lmac = &bgx->lmac[0];
-		if (lmac->lmac_type == BGX_MODE_QSGMII)
-			return 1;
+	if (bgx)
 		return bgx->lmac_count;
-	}
 
 	return 0;
 }
@@ -197,28 +189,11 @@ EXPORT_SYMBOL(bgx_set_lmac_mac);
 void bgx_lmac_rx_tx_enable(int node, int bgx_idx, int lmacid, bool enable)
 {
 	struct bgx *bgx = bgx_vnic[(node * MAX_BGX_PER_CN88XX) + bgx_idx];
-	struct lmac *lmac;
 	u64 cfg;
 
 	if (!bgx)
 		return;
 
-	/* For QSGMII one physical interface works with all 4 LMACs,
-	 * hence all LMACs on the BGX needs to be set
-	 */
-	lmac = &bgx->lmac[0];
-	if (lmac->lmac_type == BGX_MODE_QSGMII) {
-		for (lmacid = 0; lmacid < bgx->lmac_count; lmacid++) {
-			cfg = bgx_reg_read(bgx, lmacid, BGX_CMRX_CFG);
-			if (enable)
-				cfg |= CMR_PKT_RX_EN | CMR_PKT_TX_EN;
-			else
-				cfg &= ~(CMR_PKT_RX_EN | CMR_PKT_TX_EN);
-			bgx_reg_write(bgx, lmacid, BGX_CMRX_CFG, cfg);
-		}
-		return;
-	}
-
 	cfg = bgx_reg_read(bgx, lmacid, BGX_CMRX_CFG);
 	if (enable)
 		cfg |= CMR_PKT_RX_EN | CMR_PKT_TX_EN;
@@ -427,20 +402,19 @@ static int bgx_lmac_sgmii_init(struct bgx *bgx, struct lmac *lmac)
 		return -1;
 	}
 
-	/* power down, reset autoneg, autoneg enable */
-	cfg = bgx_reg_read(bgx, lmacid, BGX_GMP_PCS_MRX_CTL);
-	cfg &= ~PCS_MRX_CTL_PWR_DN;
-	cfg |= (PCS_MRX_CTL_RST_AN | PCS_MRX_CTL_AN_EN);
-	bgx_reg_write(bgx, lmacid, BGX_GMP_PCS_MRX_CTL, cfg);
-
 	if (lmac->lmac_type == BGX_MODE_QSGMII) {
 		/* Disable disparity check for QSGMII */
 		cfg = bgx_reg_read(bgx, lmacid, BGX_GMP_PCS_MISCX_CTL);
 		cfg &= ~PCS_MISC_CTL_DISP_EN;
 		bgx_reg_write(bgx, lmacid, BGX_GMP_PCS_MISCX_CTL, cfg);
-		return 0;
 	}
 
+	/* power down, reset autoneg, autoneg enable */
+	cfg = bgx_reg_read(bgx, lmacid, BGX_GMP_PCS_MRX_CTL);
+	cfg &= ~PCS_MRX_CTL_PWR_DN;
+	cfg |= (PCS_MRX_CTL_RST_AN | PCS_MRX_CTL_AN_EN);
+	bgx_reg_write(bgx, lmacid, BGX_GMP_PCS_MRX_CTL, cfg);
+
 	if (bgx_poll_reg(bgx, lmacid, BGX_GMP_PCS_MRX_STATUS,
 			 PCS_MRX_STATUS_AN_CPT, false)) {
 		dev_err(&bgx->pdev->dev, "BGX AN_CPT not completed\n");
@@ -803,11 +777,6 @@ static void bgx_init_hw(struct bgx *bgx)
 		bgx_reg_write(bgx, 0, BGX_CMR_RX_STREERING + (i * 8), 0x00);
 }
 
-static u8 bgx_get_lane2sds_cfg(struct bgx *bgx, struct lmac *lmac)
-{
-	return (u8)(bgx_reg_read(bgx, lmac->lmacid, BGX_CMRX_CFG) & 0xFF);
-}
-
 static void bgx_print_qlm_mode(struct bgx *bgx, u8 lmacid)
 {
 	struct device *dev = &bgx->pdev->dev;
@@ -848,12 +817,6 @@ static void bgx_print_qlm_mode(struct bgx *bgx, u8 lmacid)
 			dev_info(dev, "%s: 40G_KR4\n", (char *)str);
 		break;
 	case BGX_MODE_QSGMII:
-		if ((lmacid == 0) &&
-		    (bgx_get_lane2sds_cfg(bgx, lmac) != lmacid))
-			return;
-		if ((lmacid == 2) &&
-		    (bgx_get_lane2sds_cfg(bgx, lmac) == lmacid))
-			return;
 		dev_info(dev, "%s: QSGMII\n", (char *)str);
 		break;
 	case BGX_MODE_INVALID:
@@ -862,11 +825,12 @@ static void bgx_print_qlm_mode(struct bgx *bgx, u8 lmacid)
 	}
 }
 
-static void lmac_set_lane2sds(struct bgx *bgx, struct lmac *lmac)
+static void lmac_set_lane2sds(struct lmac *lmac)
 {
 	switch (lmac->lmac_type) {
 	case BGX_MODE_SGMII:
 	case BGX_MODE_XFI:
+	case BGX_MODE_QSGMII:
 		lmac->lane_to_sds = lmac->lmacid;
 		break;
 	case BGX_MODE_XAUI:
@@ -876,14 +840,6 @@ static void lmac_set_lane2sds(struct bgx *bgx, struct lmac *lmac)
 	case BGX_MODE_RXAUI:
 		lmac->lane_to_sds = (lmac->lmacid) ? 0xE : 0x4;
 		break;
-	case BGX_MODE_QSGMII:
-		/* There is no way to determine if DLM0/2 is QSGMII or
-		 * DLM1/3 is configured to QSGMII as bootloader will
-		 * configure all LMACs, so take whatever is configured
-		 * by low level firmware.
-		 */
-		lmac->lane_to_sds = bgx_get_lane2sds_cfg(bgx, lmac);
-		break;
 	default:
 		lmac->lane_to_sds = 0;
 		break;
@@ -909,7 +865,7 @@ static void bgx_set_lmac_config(struct bgx *bgx, u8 idx)
 		lmac->use_training =
 			bgx_reg_read(bgx, 0, BGX_SPUX_BR_PMD_CRTL) &
 				SPU_PMD_CRTL_TRAIN_EN;
-		lmac_set_lane2sds(bgx, lmac);
+		lmac_set_lane2sds(lmac);
 		return;
 	}
 
@@ -928,7 +884,7 @@ static void bgx_set_lmac_config(struct bgx *bgx, u8 idx)
 		lmac->use_training =
 			bgx_reg_read(bgx, idx, BGX_SPUX_BR_PMD_CRTL) &
 				SPU_PMD_CRTL_TRAIN_EN;
-		lmac_set_lane2sds(bgx, lmac);
+		lmac_set_lane2sds(lmac);
 
 		/* Set LMAC type of other lmac on same DLM i.e LMAC 1/3 */
 		olmac = &bgx->lmac[idx + 1];
@@ -936,7 +892,7 @@ static void bgx_set_lmac_config(struct bgx *bgx, u8 idx)
 		olmac->use_training =
 		bgx_reg_read(bgx, idx + 1, BGX_SPUX_BR_PMD_CRTL) &
 			SPU_PMD_CRTL_TRAIN_EN;
-		lmac_set_lane2sds(bgx, olmac);
+		lmac_set_lane2sds(olmac);
 	}
 }
 
@@ -947,7 +903,7 @@ static bool is_dlm0_in_bgx_mode(struct bgx *bgx)
 	if (!bgx->is_81xx)
 		return true;
 
-	lmac = &bgx->lmac[0];
+	lmac = &bgx->lmac[1];
 	if (lmac->lmac_type == BGX_MODE_INVALID)
 		return false;
 
@@ -973,7 +929,7 @@ static void bgx_get_qlm_mode(struct bgx *bgx)
 	if (bgx->lmac_count > MAX_LMAC_PER_BGX)
 		bgx->lmac_count = MAX_LMAC_PER_BGX;
 
-	for (idx = 0; idx < MAX_LMAC_PER_BGX; idx++)
+	for (idx = 0; idx < bgx->lmac_count; idx++)
 		bgx_set_lmac_config(bgx, idx);
 
 	if (!bgx->is_81xx) {
-- 
1.9.1

